### PR TITLE
Pull Request to fix chat messages

### PR DESCRIPTION
**Pull request by JoachimFlottorp**
My knowledge of this is these two variables return false
`std::string playerName = user->GetLastUsedChar()->GetName();`
`bool isMythran = user->GetLastUsedChar()->GetGMLevel() > 0;`

. Which in return makes the message fail to send.

image
Result if `if (!user->GetLastChatMessageApproved() && !isMythran) return;` is commented out!

- [x] Fix User is mythran
- [x] Fix Message is approved

Im not good at c++ at all, so if someone with more knowledge about c++ would like to help. that would be greatly appreciated.

### CODE REVIEW

**Review by DarwinAnim8or — CHANGES_REQUESTED** [review]
I'm confused here-- the name being fetched has nothing to do with it. 
The mythran check is to see if you can bypass the spell check / word filter.

You have incorrect spelling in your message shown in the image, which means it's denied. DLU's chat works by checking against a list of approved words, not a list of bad words to be blocked. 

Please clarify what the issue is, and make sure it's not just the spelling errors that are preventing your messages. 
Thank you.

(marking as "requesting changes" seeing as it needs further elaboration and the code itself is just adding two log calls which shouldn't be there in a production enviroment and should be done with a debugger instead in a development enviroment)

**Comment by JoachimFlottorp** [comment]
All those words i sent are verified, they show up as you type and all words can be found in the chatplus_en_us.txt file in build/res

The reason the message is red is because the server did not pass the message through the filter. Due to it being.. Broken.

i've looked further into it and found out that the problem is in file: dChatFilter.cpp `dChatFilter::IsSentenceOkay` which returns false when it checks up the messages against the filter. 
The message get's returned at the `if (!IsInWordlist(hash)`.
Even though just like i said, the message fits the filter, and should get passed through the filter.

The user i am running has a GM level of 9, which should set isMythran to true, however it does not.

**Comment by DarwinAnim8or** [comment]
Note that account and character GM levels are seperate-- the account GM level is the maximum GM level that can be applied to a character owned by that account. 

That means you'd have to do /gmlevel <level> or one of it's alias commands in order to set the GM level for your character, if you have GM enabled then you should see a green "send" button rather than the default one.

Could you verify this first?
I couldn't replicate the bug you're describing, though I have seen that sometimes the client will still show red despite it being okay, but adding and removing a space usually fixes it

**Comment by averysumner** [comment]
There are reports of chat working for other users, so there is definitely just something wrong with your configuration.

**Comment by JoachimFlottorp** [comment]
Alright, i see what you mean by gm level, i was unaware that characters could have a gm level itself. 
However i do not understand how you are able to chat. As it is happening to every server owner that i have talked to. Can you possibly also try to set one of your characters to gm level 0 and try to type yourself. 

I've been looking more on how the filtering is done, and i suspect that how the DCF files are written is "broken" or done in a wrong way. as the word, for example "test" exists in the `chatplus_en_us.txt` but does not exist inside the converted `chatplus_en_us.dcf`. However, don't take my word for it and i could be wrong. i will commit a version with some debugging, this does not have fix anything, but shows more how the hashed data looks like.

**Comment by averysumner** [comment]
Upon further inspection, the issue seems to lie with the line endings in the chatplus file. Look at #61 for more details. If you’d like you can commit a fix for this problem.

**Comment by JoachimFlottorp** [comment]
Alright, lol. Seems like that was a very easy way to fix it. Thanks to @Wincent01 and @averysumner 
I'll commit a change to the README file talking about how to fix this.

**Comment by JoachimFlottorp** [comment]
Open to change if the README does not specify well enough.

**Comment by averysumner** [comment]
I’m wondering if it might be better to change the dcf code to just use windows line endings, probably easier for people setting it up.

**Comment by Wincent01** [comment]
Maybe a better solution would be to accept both line-endings. Something along the lines of this
```c++
...
while (std::getline(file, line)) {
        line.erase(std::remove(line.begin(), line.end(), '\r'), line.end()); //Remove nix line-endings
	std::transform(line.begin(), line.end(), line.begin(), ::tolower); //Transform to lowercase
	m_Words.push_back(CalculateHash(line));
}
```

**Comment by JoachimFlottorp** [comment]
@Wincent01 yeah, sounds like a better way of doing it, since you know how to do it.

File: dNet/ClientPackets.cpp [hunk 1]
```diff
@@ -64,6 +64,8 @@ void ClientPackets::HandleChatMessage(const SystemAddress& sysAddr, Packet* pack
 
 	std::string playerName = user->GetLastUsedChar()->GetName();
 	bool isMythran = user->GetLastUsedChar()->GetGMLevel() > 0;
+	Game::logger->Log("IsUserMythran", "%s", isMythran?"Yes, they are mythran\n":"No, they are not mythran\n");
+	Game::logger->Log("IsMessageApproved", "%s", user->GetLastChatMessageApproved()?"Yes, Their message is approved\n":"No, their message is not approved\n");
```
Comment (Wincent01) [inline dNet/ClientPackets.cpp:68]: Remove these

File: README.md [hunk 2]
```diff
@@ -162,6 +162,15 @@ certutil -hashfile <file> SHA256
...
 
+**Fix chatplus_en_us.txt**
+* If the file has been made in windows it has to be converted to support unix. Inside of WSL run this command
+* Move yourself into build/res using cd if you are not yet in there.
+```bash
+cd build/res
+sudo apt install dos2unix
+dos2unix -o chatplus_en_us.txt chatplus_en_us.txt
+```
+
```
Comment (Wincent01) [inline README.md:173]: If we allow both line-endings this is unnecessary

**Comment by JoachimFlottorp** [comment]
Something like this?

File: dChatFilter/dChatFilter.cpp [hunk 3]
```diff
@@ -105,9 +105,22 @@ bool dChatFilter::IsSentenceOkay(const std::string& message, int gmLevel) {
...
+		m_Words.push_back(CalculateHash(line));
+	}
+	#else 
 	while (std::getline(sMessage, segment, ' ')) {
+	#endif
+
+
 		std::transform(segment.begin(), segment.end(), segment.begin(), ::tolower); //Transform to lowercase
-		segment = std::regex_replace(segment, reg, "");
+			segment = std::regex_replace(segment, reg, "");
```
Comment (Wincent01) [inline dChatFilter/dChatFilter.cpp]: The erase should be unconditional of platform. This will allow use of `\r\n` line-endings on Unix

Comment (JoachimFlottorp) [reply dChatFilter/dChatFilter.cpp]: New to this pull request on github. Am i the only one that can make commits?

**Review by DarwinAnim8or — APPROVED** [review]

**Pull request merged**